### ISSUE TITLE
bisecting for gitlab-ce #14972

### ISSUE THREAD

**Issue by glensc** [libgit2/libgit2#3802]
as @ethomson pointed out in https://github.com/libgit2/rugged/issues/607#issuecomment-221923851 i should be bisecting the commit that fixes [gitlab-ce #14972](https://gitlab.com/gitlab-org/gitlab-ce/issues/14972)

here will be my results bisecting
https://github.com/libgit2/libgit2/compare/v0.24.0...c75644

**Comment by glensc** [comment]
first, i was not able to bisect 0.24.0:

```
$ git bisect start v0.24.0 c756448cefdf7369f30c45e9427ad5ba67e33483
Some good revs are not ancestor of the bad rev.
git bisect cannot work properly in this case.
Maybe you mistook good and bad revs?
```

so as manually tested 0.24.1 also has the problem, will be bisecting with 0.24.1

**Comment by glensc** [comment]
v0.24.1 .. c756448cefdf7369f30c45e9427ad5ba67e33483

but not sure what i do wrong here, it runs script only once, script returns 2, and whole bisecting is aborted:

```
git bisect reset
git bisect start v0.24.1 c756448cefdf7369f30c45e9427ad5ba67e33483
git bisect run /bisect-14972.sh
...
+ exit 2
The merge base 785d8c48ea8725691da3c50e7dae8751523d4c30 is bad.
This means the bug has been fixed between 785d8c48ea8725691da3c50e7dae8751523d4c30 and [c756448cefdf7369f30c45e9427ad5ba67e33483].
bisect run failed:
'bisect_state bad' exited with error code 3
```

**Comment by carlosmn** [comment]
The releases aren't merged into the master branch. So comparing 0.24.1
against master means it would have to go backwards first to a merge base,
which doesn't sound like something that would produce reliable results.

Testing the tip of master against the latest common commit between it and
the 0.24 maintenance branch should give you an usable list of commits to
try.

**Comment by glensc** [comment]
looks like i just need  to swap good and bad arguments:
http://stackoverflow.com/a/17153598/2314626

so the run gave me this:

bf804d407e8d1fcff42e1113aa286270ae8925c0 is the first bad commit
commit bf804d407e8d1fcff42e1113aa286270ae8925c0
```
Author: Carlos Martín Nieto <carlosmn@github.com>
Date:   Thu Mar 17 10:45:22 2016 +0100

    commit: fix extraction of single-line signatures
    
    The function to extract signatures suffers from a similar bug to the
    header field finding one by having an unecessary line feed check as a
    break condition of its loop.
    
    Fix that and add a test for this single-line signature situation.

:040000 040000 c4cc5246800567d1eddd7c6f17df9f94d20007f5 79b01b32889cda17ef9269349da9aa55d2a1ea52 M      src
:040000 040000 9792a8c6616e731aba48ba464aa68fecdac80806 d8b5c44450a4cae89cbc8375b76f74ac972531b5 M      tests
bisect run success
[glen@blodnatt libgit2]$ 
```


i will be rechecking surrounding commits manually shortly. and whether the swapping of arguments was right thing to do

**Comment by glensc** [comment]
full(er) output showing it can still bisect:

```
[glen@blodnatt libgit2]$ git bisect start --term-new=fixed --term-old=unfixed
[glen@blodnatt libgit2]$ git bisect fixed c756448cefdf7369f30c45e9427ad5ba67e33483
[glen@blodnatt libgit2]$ git bisect unfixed v0.24.0
Bisecting: 129 revisions left to test after this (roughly 7 steps)
[df25daef9b07174124989aaf993cc20c9e9b0660] Added clar test for #3568
[glen@blodnatt libgit2]$ 
```

**Comment by glensc** [comment]
it gave me this:
```
...
c756448cefdf7369f30c45e9427ad5ba67e33483 is the first fixed commit
bisect run success
[glen@blodnatt libgit2]$ git for-each-ref refs/bisect
c756448cefdf7369f30c45e9427ad5ba67e33483 commit refs/bisect/fixed
17a93afa782ac8aebb37ecb98b928215f36e2008 commit refs/bisect/unfixed-17a93afa782ac8aebb37ecb98b928215f36e2008
2f3f1ee08532b10993b7e9270e5fe61f08f795ee commit refs/bisect/unfixed-2f3f1ee08532b10993b7e9270e5fe61f08f795ee
6ee08d2cd0671d42e4148988863a6a40fbe721cb commit refs/bisect/unfixed-6ee08d2cd0671d42e4148988863a6a40fbe721cb
785d8c48ea8725691da3c50e7dae8751523d4c30 commit refs/bisect/unfixed-785d8c48ea8725691da3c50e7dae8751523d4c30
88284dfb7905c5990babb4238b7cd30bdf823500 commit refs/bisect/unfixed-88284dfb7905c5990babb4238b7cd30bdf823500
a2cb47130ec7662811fe3447f69bae3f176e0362 commit refs/bisect/unfixed-a2cb47130ec7662811fe3447f69bae3f176e0362
c148533024319689fca016f7c452d556265bb13f commit refs/bisect/unfixed-c148533024319689fca016f7c452d556265bb13f
c91a1dc1d975aa32c58cfafbf1625dcbb0be7c3c commit refs/bisect/unfixed-c91a1dc1d975aa32c58cfafbf1625dcbb0be7c3c
df25daef9b07174124989aaf993cc20c9e9b0660 commit refs/bisect/unfixed-df25daef9b07174124989aaf993cc20c9e9b0660
[glen@blodnatt libgit2]$ 
```

meh? (still n00b at bisecting)

**Comment by glensc** [comment]
found the commit that fixed for me:
https://github.com/libgit2/libgit2/commit/be30387

**Comment by glensc** [comment]
note to future self, as the bisecting "started" with "good" by default, while git bisecting was initially written to find commit that "broke", but i was searching for the commit that "fixed" so therefore needed to swap good vs bad logic, arguments and the exit code in script to run

```
$ git bisect reset
$ git bisect start --term-new=fixed --term-old=unfixed c756448cefdf7369f30c45e9427ad5ba67e33483 v0.24.0
$ git bisect run /bisect-14972.sh
``

**Comment by glensc** [comment]
So, what next? Can you backport this to 0.24.2? Have you figured out how to create regression test based on the fix (my testcases are all in private repos)?

**Comment by glensc** [comment]
btw, the repos which i have problems with are generated with [agito](https://github.com/fragglet/agito) which uses [dulwich](http://dulwich.readthedocs.io/en/latest/) underneath.

my first imports were created with dulwich 0.11.2, i've re-tested with recent dulwich 0.13.0, the problem remains.

not sure is this helpful

**Comment by ethomson** [comment]
I think that the current status is:

1. Your bug will be fixed by libgit2 0.25.  We're nearing a release, but we're not done yet.
2. Since your issue was fixed by very severe changes (basically a rewrite of the iterators), it's not something that we'd be able to backport to 0.24.  We like those to be quite stable and - although we'd like to fix an iterator bug, I wouldn't feel comfortable doing it by pulling in such a massive change, especially not knowing what the bug actually was.
3. There's already a rugged release out with this change.  Since gitlab depends on rugged, _not_ libgit2 directly, this seems like an appropriate release for them to target.
4. If they really want a libgit2 0.24.x that includes this change, they're welcome to backport this themselves.

I don't think there's a lot actionable to do in libgit2 at the moment.  Thanks for reporting this - if you're able to provide a repo that illustrates this behavior, that would be great.

For example, does `git fast-export --anonymize` produce something that is suitably anonymous?

**Comment by glensc** [comment]
gitlab dudes said they want stable rugged release, which 0.25.0b3 is not ... so have to wait a bit longer

**Comment by glensc** [comment]
as for git-fast-export, the problem is not visible if i export and import, so i don't think export would help. i think git-fast-import creates the repo differently which older libgit2 reads normally.

**Comment by glensc** [comment]
looks like the problem isn't really solved. found another case where it shows wrong file count using same rugged library that solved previous issues i had.
